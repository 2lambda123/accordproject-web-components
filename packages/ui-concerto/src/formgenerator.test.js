/*
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
import React from 'react';
import { ModelManager } from '@accordproject/concerto-core';
import { render, waitFor, screen } from '@testing-library/react';
import Generator from './formgenerator';

const fs = require('fs');

describe('formgenerator Tests', () => {
  const modelManager = new ModelManager();
  modelManager.addModelFile(
    `namespace org.accordproject.base
  abstract asset Asset {  }
  abstract participant Participant {  }
  abstract transaction Transaction identified by transactionId {
    o String transactionId
  }
  abstract event Event identified by eventId {
    o String eventId
  }`,
    'org.accordproject.base.cto',
    false,
  );
  modelManager.updateExternalModels();
  describe('#validation', () => {
    it('accepts a model string as input', async () => {
      const generator = new Generator(modelManager);
      expect(generator).not.toBeNull();
    });
  });
  describe('#instantiate', () => {
    it('generates a form from text', async () => {
      const text = fs.readFileSync(`${__dirname}/../test/data/relationship.cto`, 'utf8');
      const options = {
        customClasses: {
          field: 'form-group',
          input: 'form-control',
          label: 'control-label'
        },
        updateExternalModels: true,
      };
      modelManager.addModelFile(text, 'model', true);
      modelManager.updateExternalModels();
      const generator = new Generator(modelManager, options);
      expect(generator).not.toBeNull();
      expect(generator.getTypes()).toHaveLength(2);

      const generatedInstance = generator.generateJSON('test.Person');
      const json = {
        ...generatedInstance,
        address: {
          ...generatedInstance.address,
          $identifier: 'a9c0177b-f08f-417e-a956-7a7c9e9701e5',
        }
      };
      const form = generator.generateHTML('test.Person', json);

      const { container } = render(<div>{ form }</div>);
      await waitFor(() => screen.getByText('Name'));
      expect(container).toMatchSnapshot();
    });
  });
});
