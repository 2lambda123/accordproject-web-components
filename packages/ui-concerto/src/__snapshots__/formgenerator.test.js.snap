// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`formgenerator Tests #instantiate generates a form from text 1`] = `
<div>
  <div>
    <div
      class="required field"
    >
      <label
        for="name"
      >
        Name
      </label>
      <div
        class="ui input"
      >
        <input
          id="name"
          type="text"
          value="resource1"
        />
      </div>
    </div>
    <div
      class="required field"
    >
      <label>
        Address
      </label>
      <div
        class="required field"
      >
        <label
          for="address.$identifier"
        >
          $identifier
        </label>
        <div
          class="ui input"
        >
          <input
            id="address.$identifier"
            type="text"
            value="a9c0177b-f08f-417e-a956-7a7c9e9701e5"
          />
        </div>
      </div>
      <div
        class="required field"
      >
        <label
          for="address.street"
        >
          Street
        </label>
        <div
          class="ui input"
        >
          <input
            id="address.street"
            type="text"
            value=""
          />
        </div>
      </div>
      <div
        class="required field"
      >
        <label
          for="address.city"
        >
          City
        </label>
        <div
          class="ui input"
        >
          <input
            id="address.city"
            type="text"
            value=""
          />
        </div>
      </div>
      <div
        class="required field"
      >
        <label>
          Country
        </label>
        <div
          aria-expanded="false"
          class="ui fluid selection dropdown"
          role="listbox"
          tabindex="0"
        >
          <div
            aria-atomic="true"
            aria-live="polite"
            class="text"
            role="alert"
          >
            USA
          </div>
          <i
            aria-hidden="true"
            class="dropdown icon"
          />
          <div
            class="menu transition"
          >
            <div
              aria-checked="true"
              aria-selected="true"
              class="active selected item"
              role="option"
              style="pointer-events: all;"
            >
              <span
                class="text"
              >
                USA
              </span>
            </div>
            <div
              aria-checked="false"
              aria-selected="false"
              class="item"
              role="option"
              style="pointer-events: all;"
            >
              <span
                class="text"
              >
                UK
              </span>
            </div>
            <div
              aria-checked="false"
              aria-selected="false"
              class="item"
              role="option"
              style="pointer-events: all;"
            >
              <span
                class="text"
              >
                France
              </span>
            </div>
            <div
              aria-checked="false"
              aria-selected="false"
              class="item"
              role="option"
              style="pointer-events: all;"
            >
              <span
                class="text"
              >
                Sweden
              </span>
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      class="field"
    >
      <label>
        Children
      </label>
      <div
        class="arrayElement grid"
      >
        <div />
        <div>
          <button
            aria-label="Add an element to Children"
            class="ui basic circular icon button arrayButton"
          >
            <i
              aria-hidden="true"
              class="plus circle icon"
            />
          </button>
        </div>
      </div>
    </div>
  </div>
</div>
`;
